src/WebJobs.Script/IO/AutoRecoveringFileSystemWatcher.cs

```diff
@@ -0,0 +1,202 @@
+            }
+        }
+
+        private void AddEventSubscriptions(FileSystemWatcher fileWatcher, WatcherChangeTypes changeTypes)
+        {
+            if (changeTypes.HasFlag(WatcherChangeTypes.Changed))
+            {
+                fileWatcher.Changed += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Created))
+            {
+                fileWatcher.Created += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Deleted))
+            {
+                fileWatcher.Deleted += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Renamed))
+            {
+                fileWatcher.Renamed += OnFileChanged;
+            }
+        }
+
+        private void RemoveEventSubscriptions(FileSystemWatcher fileWatcher)
+        {
+            fileWatcher.Changed -= OnFileChanged;
+            fileWatcher.Created -= OnFileChanged;
+            fileWatcher.Deleted -= OnFileChanged;
+            fileWatcher.Renamed -= OnFileChanged;
+        }
+
+        protected void OnFileChanged(object sender, FileSystemEventArgs e)
+        {
+            Changed?.Invoke(this, e);
+        }
+
+        protected void OnFileWatcherError(ErrorEventArgs obj)
+        {
+            Trace($"Failure detected '{obj.GetExceptionMessage()}'. Initiating recovery...", TraceLevel.Warning);
+
+            Recover().ContinueWith(t =>
+            {
+                if (t.IsFaulted || t.IsCanceled)
+                {
+                    t.Exception?.Handle(e => true);
+                    Trace($"Recovery process aborted.", TraceLevel.Error);
+                }
+                else
+                {
+                    Trace("File watcher recovered.", TraceLevel.Info);
+                }
+            });
+        }
+
+        private void Trace(string message, TraceLevel level)
+        {
+            _traceWriter?.Trace($"File watcher: ('{_path}') - {message}", level, null);
+        }
+
+        private async Task Recover(int attempt = 1)
+        {
+            // Exponential backoff on retries
+            long wait = Convert.ToInt64((Math.Pow(2, attempt) - 1) / 2);
+            
+            // maximum wait of 5 minutes
+            wait =  Math.Min(wait, 300);
+
+            if (wait > 0)
+            {
+                Trace($"Next recovery attempt in {wait} seconds...", TraceLevel.Warning);
+                await Task.Delay(TimeSpan.FromSeconds(wait)).ConfigureAwait(false);
+            }
+
+            // Check if cancellation was requested while we were waiting
+            _cancellationToken.ThrowIfCancellationRequested();
+
+            try
+            {
+                ReleaseCurrentFileWatcher();
+                InitializeWatcher();
+            }
+            catch (Exception exc) when (!(exc is TaskCanceledException))
+            {
+                Trace($"Unable to recover - {exc.Message}", TraceLevel.Error);
```

Comment:
It may be useful to have the stack in this message -- can you use `exc.ToString()`?